### ISSUE TITLE
Google PageSpeed score

### ISSUE THREAD

**Issue by myalban** [maplibre/maplibre-gl-js#1876]
## Motivation

Is it possible to improve the JS parsing time ?
Google page speed test is bellow 50%  for smartphone.

[https://pagespeed.web.dev/report?url=https%3A%2F%2Fmaplibre.org%2Fmaplibre-gl-js-docs%2Fexample%2Fdisable-rotation%2F](
https://pagespeed.web.dev/report?url=https%3A%2F%2Fmaplibre.org%2Fmaplibre-gl-js-docs%2Fexample%2Fdisable-rotation%2F)

## Design Alternatives

Split package  with limited functions ?

**Comment by HarelM** [comment]
One of the main problems related to this code base is that it uses workers.
In order to simplify the usage of this library the workers are rigged into the library code.
This creates a huge problem for splitting, tree shaking etc.
Feel free to research this and propose a solution.
See also #977

**Comment by rotu** [comment]
I think this is fairly easy. Using the CSP build instead of the main one with custom Javascript loaders sped things up significantly. Also allows us to use modulepreload to get the worker to start loading before the main script is even started.

[lighthouse-before.pdf](https://github.com/maplibre/maplibre-gl-js/files/10338327/lighthouse-before.pdf)
[lighthouse-after.pdf](https://github.com/maplibre/maplibre-gl-js/files/10338329/lighthouse-after.pdf)

I suggest we just replace the main build with the CSP one!

**Comment by HarelM** [comment]
Replacing the main with the CSP and shifting the configuration to the developer is problematic.
If CSP can solve performance issue I suggest to better document it, but I don't think it should be the main build.
Unless I'm missing something obvious here, when I looked at the CSP docs I got scared and decided to avoid it, might be just me though...

**Comment by rotu** [comment]
CSP itself is not solving performance issues - I think that wrangling the script into a blob is *creating* performance issues. (The script can't be parsed ahead of time and still has to be parsed in both the main and worker thread anyway). It just so happens that the blob approach *also* causes issues with CSP. Also, it means the worker script cannot be stored in browser cache. (in my above numbers, the cache is disabled anyway, so I think the new approach is winning even fetching the worker script 4 times!)

Additionally, CSP should not shift configuration to the developer - this package should definitely run "out of the box" without specifying a worker URL. I'm thinking that we should just set the worker URL to a sane default (e.g. `new URL('./worker.js', import.meta.url)` and release the worker script in the dist folder.

It's still necessary to have a non-esm-compatible worker script (because Firefox does not support ESM in workers).

**Comment by rotu** [comment]
Fyi @birkskyum.

**Comment by birkskyum** [comment]
Regarding firefox, there has been done more work on it in the past half year, than in the preceding 7 years [link](https://bugzilla.mozilla.org/show_bug.cgi?id=1247687), so I so we can finally have this.

I would love to see a demonstration of this work well with ESM only in say Chrome or safari to validate that our path ahead is clear.

**Comment by clementmas** [comment]
In terms of "time to first paint", I find that loading the local GeoJSON first before the base style makes it seem much faster.
It might not improve performance. Or possibly make it worse since the GeoJSON has to be repainted over the base style once loaded. But that could be something to explore.

**Comment by Canoir** [comment]
hi, I don't actully understand the CSP build? what is that? i have such a problem too with greate core web vitals but 10 sec and 7 sec TBT and Speed index. without map everything is good.

**Comment by rotu** [comment]
Let’s start with “CSP”. It’s security model based around restricting where executable code can come from. Certain CSP options can forbid running code from a string via `eval` or from a `blob:` URL (aka object url).

Now in order to keep the size of this library down, some of the code which is used in both the main process and background workers is essentially kept as a string until late in the game and then constructs the worker code around it *after* the library has loaded. This is slow (and I think also prevents certain optimizations even when the page is already cached).

The CSP build does not do this - it ships the main library and the worker code as two separate scripts, duplicating the code but (I believe) running much faster.

I’ll see if I can put some solid measurements to my above wishy-washy description and speculation.

**Comment by rotu** [comment]
UPDATE: I can't seem to find the performance discrepancy I expected. Then again I'm testing on the `test/examples/` pages, not the previous debug pages. Is this issue maybe resolved?

**Comment by Canoir** [comment]
But still my website has so much TBT and speed index because of maplibre-gl. website address is : https://behtarino.com

**Comment by rotu** [comment]
I couldn't tell ya. I ran the performance profiling on the latest version.

One thing that *could* be an impact is that your site is using multiple workers and maplibre uses a single worker as of https://github.com/maplibre/maplibre-gl-js/pull/2354

Another thing is that some web requests are curiously slow and have caching disabled (e.g. `behtarino-manifest.json` seems to take like 600 ms for some reason)

**Comment by CommanderStorm** [comment]
At least looking at https://pagespeed.web.dev/analysis/https-maplibre-org-maplibre-gl-js-docs-example-disable-rotation/l6byiij5x5?form_factor=mobile this issue seems to be resolved.

There is still a lot of ground to cover (there are a few other issues related to performance if people are curious), but progress towards the goal.

**Comment by jasongitmail** [comment]
It appears the page under test on that link is a 404. just a heads up in case of invalid results due to that

<img width="1035" height="708" alt="Image" src="https://github.com/user-attachments/assets/bf1c33ed-5df0-4311-8e9c-45ae4ef5edb1" />

**Comment by CommanderStorm** [comment]
Here is the updated link, no clue when this broke.

https://pagespeed.web.dev/analysis/https-maplibre-org-maplibre-gl-js-docs-examples-disable-map-rotation/mg6m2nqkf1?form_factor=mobile

**Comment by martinfrancois** [comment]
Still a problem, and I have numbers from a production site rather than from the
docs example.

**Setup.** A personal site with one map: MapLibre 6.3.0 through react-map-gl
8.1.2, CARTO Positron vector basemap, country polygons plus city points, popups,
a navigation control. The map already has every cheap option turned on:
`renderWorldCopies: false`, `fadeDuration: 0`, `dragRotate: false`,
`maxPitch: 0`, `maxZoom: 10`, `attributionControl: false`, and the GeoJSON is
handed to the worker as a URL so the main thread never parses it. The page waits
for LCP before it creates the map.

**How I measured.** Chrome, mobile viewport 412x823, 4x CPU throttle (the same
throttle PageSpeed uses for mobile), CDP CPU profile at 100 microsecond sampling,
main thread only. Numbers are medians of repeated runs.

### Where the time goes

MapLibre's own code costs **770 ms of main-thread time**. The split is the
interesting part:

| | time | share |
| --- | --- | --- |
| compile + evaluate the bundle | **147 ms** | 19% |
| construct, style and render the map | **623 ms** | 81% |

I got the 147 ms from `EvaluateScript` for the MapLibre chunk in a
`devtools.timeline` trace, and the 770 ms from summing self time per script in
the CPU profile.

### There is no hot spot

This is what surprised me. Self time per function, biggest first:

```
receive           20 ms
_render           13 ms
_setupPainter     11 ms
nO                11 ms
render            10 ms
_updateCompact     8 ms
_updatePlacement   7 ms
...and a long tail below 7 ms
```

The largest single function is 20 ms out of 770 ms. It is not one slow thing. It
is a few hundred small things, which is what a map library looks like when it is
doing its job.

### What this means for "split package with limited functions?"

The original post asked whether a smaller build would fix this, and #3708 asks
for the same. My numbers say: **it would help less than people expect.**

A slimmer build can only touch the 147 ms of parsing and evaluating. That is 19%.
The other 81% is style parsing, layer construction, painter setup, symbol
placement, and drawing frames, and none of that goes away when you remove terrain
or globe support from the bundle. The shipped bundle really does carry terrain,
globe, hillshade, heatmap, sky, atmosphere, video and RTL text code that my map
never uses, so there is genuine weight to remove. It just is not where the time
is.

I am not arguing against a modular build. I am saying that if someone builds one
and expects the PageSpeed score to jump, they will be disappointed, and it would
be good to know that before the work starts.

### Things I tried that did not work, so nobody repeats them

I went through the option surface carefully before concluding this. Each of these
was checked against the 6.3.0 source, not guessed:

- **`fadeDuration: 0`** is not a load-time lever. `Style._updatePlacement` does
  `forceFullPlacement ||= this._layerOrderChanged || fadeDuration === 0`, so
  setting it to 0 actually forces full placement.
- **`{validate: false}` on `addSource` / `addLayer`** does not skip the expensive
  part. `Style._validate` builds its argument object, including
  `style: this.serialize()`, before it looks at the flag.
- **`crossSourceCollisions: false`** is slower, not faster, and it lets your
  labels collide with the basemap's.
- **`localIdeographFontFamily`, `collectResourceTiming`, `refreshExpiredTiles`,
  `maxTileCacheSize`, `trackResize`, `workerCount`, `pixelRatio`,
  `maplibreLogo`** are already at their cheapest value, already the default, or
  have no effect at load time.
- **`maxZoom` / `minZoom` / `maxBounds`** do not reduce tile work. `maxZoom: 10`
  does not change the covering-tile zoom.
- **Handler options** (`dragRotate`, `pitchWithRotate`, `touchPitch`, `boxZoom`,
  `keyboard`, `doubleClickZoom`, `scrollZoom`) do not avoid handler setup.
- **`reuseMaps`** in react-map-gl does nothing on a cold page load, which is
  exactly what PageSpeed measures.
- **Splitting the bundle's evaluation across tasks** is not possible from outside
  the library. One chunk is one script evaluation task.

### One real piece of waste I did find

`Style._load` runs `validateStyleMin` over the whole stylesheet synchronously
(`style/style.ts:472`). For the CARTO Positron style that is roughly 130 layers,
and validation compiles every expression through `createPropertyExpression` and
then throws the result away, immediately before `_createLayers` parses the same
expressions again.

`validateStyle: false` skips it. I measured it on my site: the MapLibre chunk went
from 779 ms to 735 ms median, but the run-to-run spread was about 50 ms and the
minimums were identical (732 vs 731 ms), so I cannot honestly call that a real
win. I mention it because doing the same expensive work twice looks like
something worth fixing inside the library, where it could be cheap, even though
turning it off from outside is not worth the loss of validation.

### Why this matters for the score, in case it is useful

PageSpeed's Total Blocking Time only counts the part of each task that goes over
50 ms. One 400 ms task costs 350 ms; eight 50 ms tasks cost nothing. So work that
is spread into small pieces is nearly free, and work that lands in one long task
is expensive. That is why "make it smaller" and "make it later" both disappoint
here, and why breaking the long tasks up would help more than either. Deferring in
particular does nothing at all: the task still lands inside the measured window.

Happy to share the raw profiles or re-run anything if it would help.

---

**AI assistance disclosure**, per CONTRIBUTING.md. This comment is the result of a
profiling session I ran with an AI coding assistant (Claude Code). It did the
profiling and drafted the write-up; I directed the work, I checked the source
citations above against maplibre-gl 6.3.0 myself, and I own what is written here.
No code is proposed, so there is nothing here to review for copyright. Every
number comes from a run against my own production site, not from a model.

**Comment by CommanderStorm** [comment]
Bigger blocker is sequential requests (js-bundle -> maplibre style-json -> tilejson -> MVT tiles) and maybe moving some code to be loaded async, now that we have updated our ECMA version to support that in maplibre@v6. Needs a lot of care though.

Our style spec is a fairly large chunk of code to download. Downloading code is usually also time spend.
So makign this better tree-shakable if `validateStyle: false` still has an effect, and one that I think you did not quite investigate, or?

Essenitally improvements are possible, but needs dedicated work if someone wants us to improve here.

**Comment by hheexx** [comment]
I have maplibre on public site loading around 1000 points with clustering and in google search console we are green on all web vitals.

**Comment by HarelM** [comment]
It's worth mentioning that the original post was pre-version 6, so the new version requires to use the old "CSP" style import which reduces the size that needs to be loaded on initial interaction.
Nevertheless, there's no doubt this can be improved if someone is willing to spend time into it.

**Comment by hheexx** [comment]
Yea, things can always be improved. Point of my post is that current situation is not bad.
Probably splitting long tasks makes most sense.

**Comment by cherenkov** [comment]
I profiled the initial load with a sampling profiler (headless Chrome on a real GPU, `--use-angle=metal`, dev build with source maps so functions resolve to files) to see what the main thread actually spends the first ~1s on, since the discussion above is mostly about bundle size. Two findings, one shipped as a small PR each and one that needs a design decision.

**1. Module evaluation: `Intl.Segmenter` (#8337).** `graphemes.ts` built two `Intl.Segmenter` at import; the first one initializes ICU (~8ms on an M1, ~35ms at 4x throttling). Only the workers shape text, so the main thread paid for it and never used it. Now lazy. `import()` median 34.6→27.1ms (1x) and 84→50ms (4x).

**2. First render: shader compilation.** With a raster + 2 line + circle + symbol style (7 programs) the `Program` constructor is the largest single cost inside the first frame on this machine, and on mobile GPUs shader compilation is much slower than on an M1. Two things make it worse than it needs to be:

- `Program` reads `COMPILE_STATUS` right after each `compileShader`, which forces the driver to finish that shader before the next one is even issued. Linking first and only checking `LINK_STATUS` (#8338) cuts the 7 programs from 18–23ms to 13ms of main-thread time.
- Programs are created lazily inside the draw calls, one at a time, so nothing overlaps. The same 7 shader pairs compiled with all compiles + links issued first and statuses/locations queried afterwards take 6ms, and with `KHR_parallel_shader_compile` (Chrome 76+, Safari 14.1+, not Firefox) polled across frames the main thread spends 0.4ms while the wall time (9–21ms) overlaps with tile loading.

### Proposal for the second part

Split `Program` into two phases and warm the cache before the first frame:

1. `Program` constructor issues `compileShader` × 2 + `linkProgram` and returns. A `finalize()` step reads `LINK_STATUS`, `applyUBOBindings`, attribute and uniform locations. `Painter.useProgram` calls `finalize()` on the cached program before handing it to a draw, so draw code does not change.
2. `Painter.prewarmPrograms(style)` runs after `style.load` (and after `addLayer`) and constructs the programs the style will need: per layer type a small table maps to program names (`background`/`backgroundPattern`, `raster`, `line`/`lineGradient`/`linePattern`/`lineSDF`, `fill` + `fillOutline`, `circle`, `symbolSDF`/`symbolIcon`, `heatmap` + `heatmapTexture`, …) plus the always-used `clippingMask`, and `sky`/`terrain` when the style has them. The cache key needs the `ProgramConfiguration.cacheKey`, which depends only on which paint properties are constant / data-driven / zoom-and-data-driven, so `new ProgramConfiguration(layer, zoom, filter)` on the style layer gives the same key the bucket's configuration will produce. A variant the table misses (e.g. `iconsInText`) just falls back to today's lazy path.
3. Where `KHR_parallel_shader_compile` exists, `finalize()` is free by the time tiles arrive; where it does not, the driver still overlaps the compiles because nothing blocks on them in between.

Expected effect on the first frame: the per-program compile stalls leave the critical path entirely, which on this machine is ~20ms per map and on mobile GPUs likely a multiple of that. The riskier piece is the layer-type → program table drifting from the draw functions; I would keep it small and treat misses as harmless.

Happy to implement it if the maintainers are fine with the `Program` split and a prewarm table in `Painter`. The profiling scripts (CDP sampling profile attributed per long task, compile-strategy microbenchmark using the captured real shader sources) are ad-hoc; I can turn the latter into an e2e benchmark under `test/bench/e2e/` if that would help reviewing.

**Comment by HarelM** [comment]
@cherenkov it's not clear what this redesign will actually improve in terms of performance. 
@martinfrancois shared a very good table describing the long tail of small methods that takes time when initializing a map.
It's probably worth trying to improve where there is most chances to reduce load time. 
While I'm not against improving performance, it's almost always the equation of maintainability of the code vs improve performance, so I'd to focus on places where either the performance gain is substantial or the changes to the code are small (like the lazy segmenter, where it's a simple fix to gain small performance).

**Comment by martinfrancois** [comment]
I tested several approaches discussed in this thread, along with a few additional ideas, to see whether they improve Google's mobile PageSpeed score. I started with local Lighthouse tests, then tested changes through PageSpeed Insights using copies of the [Disable map rotation example](https://maplibre.org/maplibre-gl-js/docs/examples/disable-map-rotation/).

Each PageSpeed comparison tests one change against an unchanged copy of the same page, with five runs per version. Scores below are **median before → after**; higher is better. I repeated comparisons that showed a higher median score. Each repeat has its own unchanged baseline. "Gain not reproduced" means the higher median score from the first test did not appear in the repeat test.

| Approach | Why I tested it | First test | Repeat tests | Conclusion |
| --- | --- | --- | --- | --- |
| Prepare shaders while tiles load, following @cherenkov's [proposal](https://github.com/maplibre/maplibre-gl-js/issues/1876#issuecomment-5557425876). | Reduce shader work during the first draws. | 45 → 44 | Not run | **No clear gain** |
| Draw once into a hidden 1×1 target. | Move first-use drawing work before the visible frame, with very few pixels to draw. | 62 → 52 | Not run | **No clear gain** |
| Draw once into a hidden full-size target. | Prepare drawing at the normal canvas size before the visible frame. | 44 → 48 | Test 2: 44 → 48<br>Test 3: 56 → 48 | **Gain not sustained** |
| Skip fully transparent sky drawing. | Avoid setup and drawing that adds no visible color. | 41 → 50 | 51 → 48 | **Gain not reproduced** |
| Preload the style JSON, following @CommanderStorm's [request-chain explanation](https://github.com/maplibre/maplibre-gl-js/issues/1876#issuecomment-5382641824). | Start the style request earlier. | 45 → 47 | 48 → 47 | **Gain not reproduced** |
| Disable fades with `fadeDuration: 0`, revisiting [my earlier comment](https://github.com/maplibre/maplibre-gl-js/issues/1876#issuecomment-5382550515). | Avoid animation frames during loading. | 42 → 52 | 45 → 39 | **Gain not reproduced** |
| Delay tile repaints by up to 100 ms. | Group tile updates into fewer repaints. | 44 → 51 | 51 → 45 | **Gain not reproduced** |

The shader row addresses @HarelM's [question about whether @cherenkov's proposal justifies its added complexity](https://github.com/maplibre/maplibre-gl-js/issues/1876#issuecomment-5557480412). The other rows test separate approaches. The details below include blocking times and individual scores, including results that differed within the same set.

The full-size off-screen draw had higher median scores in 2 of 3 comparisons, but the latest comparison scored 56 unchanged and 48 with the draw. Median blocking time was lower with the draw in all three comparisons, but the overall score gain was not sustained. The tests do not establish a reliable PageSpeed improvement from the changes above. That does not show that MapLibre cannot improve, or that another implementation or map style would give the same results.

The details below preserve the individual measurements, explain the prototypes and earlier experiments, and give the files and commands needed to check the results. The approach table uses separate tests for each change; the older combined experiments remain below as supporting history.

Each standalone table below shows all five runs and their median. 🟢 means better than the unchanged library in the same round, 🔴 means worse, and ⚪ means equal. Median markers compare the two medians. Each median is calculated separately from its row, so more green rounds do not necessarily mean a higher median score. Blocking times are rounded to the nearest millisecond. The markers describe these runs; they do not establish that the difference will repeat.

Higher PageSpeed scores are better. Lower Total Blocking Time, or TBT, is better. TBT measures blocking from long main-thread tasks during loading and contributes to the score. Other loading metrics also contribute, so lower TBT does not always produce a higher score.

Keep each comparison with its own baseline. Run order alternated between unchanged and changed versions in the new comparisons. The first sky comparison comes from the earlier five-round series, which rotated several variants. The new sky repeat uses the same alternating order as the other new comparisons. The standalone tables compare one change with the unchanged library; the older combined tests are kept in a separate details block.

<details>
<summary>Skip fully transparent sky drawing: every hosted run</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 64 | 53 | 41 | 40 | 40 | 41 |
| After: one change | 🔴 43 | 🟢 58 | 🔴 37 | 🟢 50 | 🟢 51 | 🟢 50 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,245 | 480 | 1,212 | 1,265 | 1,266 | 1,245 |
| After: one change | 🟢 1,011 | 🟢 320 | 🔴 1,788 | 🟢 624 | 🟢 560 | 🟢 624 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-encountered-pest-category-northeast-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/spi95ei5gz?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-encountered-pest-category-northeast-trycloudflare-com-current-sky-maplibre-gl-js-docs-examples-disable-map-rotation/qdxgirpy3j?form_factor=mobile).

</details>

<details>
<summary>Skip fully transparent sky drawing, repeat</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 41 | 51 | 72 | 57 | 42 | 51 |
| After: one change | 🟢 48 | 🟢 53 | 🔴 38 | 🟢 87 | 🟢 46 | 🔴 48 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,181 | 690 | 732 | 364 | 1,055 | 732 |
| After: one change | 🟢 695 | 🟢 589 | 🔴 1,553 | 🟢 212 | 🟢 1,010 | 🟢 695 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/jjyz1wxmcu?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-sky-maplibre-gl-js-docs-examples-disable-map-rotation/lnvcbzs578?form_factor=mobile).

</details>

<details>
<summary>Prepare shaders while tiles load: every hosted run</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 43 | 38 | 53 | 53 | 45 | 45 |
| After: one change | ⚪ 43 | 🟢 42 | 🟢 55 | 🔴 44 | 🟢 52 | 🔴 44 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,014 | 1,666 | 5,134 | 613 | 895 | 1,014 |
| After: one change | 🔴 1,049 | 🟢 1,031 | 🟢 3,181 | 🔴 1,130 | 🟢 536 | 🔴 1,049 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/56oi72rtrt?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-prewarm-maplibre-gl-js-docs-examples-disable-map-rotation/k0lq29ddcp?form_factor=mobile).

</details>

<details>
<summary>How the initial-load warm-up draw tests work</summary>

These tests ask whether doing the first drawing work in a hidden frame reduces blocking during the next visible frame. That extra work also costs time, so a shorter visible frame alone would not establish a PageSpeed improvement.

I tested two separate prototypes based on unchanged MapLibre 6.10.0. One draws into a 1 by 1 target; the other draws into a target matching the canvas size. Each waits until a visible tile contains drawable data, renders one hidden frame using the normal drawing code, submits the work with `gl.flush()`, and schedules the visible draw for the next animation frame. The map's render, load, and idle events are not fired for the hidden frame. Neither prototype includes the shader prewarm, sky, fade, batching, or preload changes.

1 by 1 draw: higher scores in 2 rounds, lower in 3, equal in 0. Full-size draw: higher scores in 2 rounds, lower in 2, equal in 1. Full-size draw, second test: higher scores in 4 rounds, lower in 1, equal in 0. Full-size draw, third test: higher scores in 2 rounds, lower in 2, equal in 1.

Both map screenshots matched the unchanged map, with 15 rendered features and no map or page errors. A separate diagnostic check confirmed the hidden draws used the intended target sizes and used programs that were also used by subsequent visible draws. This verifies that the code ran; it does not prove the driver reused every internal rendering pipeline. Both builds passed TypeScript checking and 110 existing painter, map-rendering, and map-event tests each.

After two higher medians, I ran a third comparison and planned a fourth only if the third median was higher. The third median was not higher, so I did not run the fourth.

The older warm-up experiments below cleared the program cache on an already loaded map. These new comparisons measure initial loading through PageSpeed, so their score results apply to these new prototypes. All changed reports include timing marks confirming that the hidden draw finished before the map's load event.

[pagespeed-draw-evidence.tar.gz](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/pagespeed-draw-evidence.tar.gz) contains the exact builds, independent patches, raw reports, scripts, and [DRAW-REPRODUCE.md](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/DRAW-REPRODUCE.md). This archive is self-contained for the two draw experiments. The optional [pagespeed-draw-screenshots-1.tar.gz](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/pagespeed-draw-screenshots-1.tar.gz) archive contains full screenshots of the PageSpeed reports. After extracting it, the commands are:

```bash
cd draw-evidence
python3 verify-draws.py
npm ci
npx playwright install chromium
python3 public-server-draws.py
```

Expose port 8140 through a public HTTPS host as described in [DRAW-REPRODUCE.md](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/DRAW-REPRODUCE.md), then run each comparison separately:

```bash
node psi-series.mjs PUBLIC_ORIGIN current,current-warmdraw 5 verify-warmdraw
python3 summarize-hosted.py verify-warmdraw
node psi-series.mjs PUBLIC_ORIGIN current,current-offscreen 5 verify-offscreen
python3 summarize-hosted.py verify-offscreen
```

Repeat a higher median score with a fresh prefix and its own unchanged baseline. The archive includes the commands for rendering checks and rebuilding both prototypes from source.

</details>

<details>
<summary>When an off-screen draw helps the PageSpeed score</summary>

The intended benefit is during initial map loading. The prototype puts first-use drawing work into a hidden frame, then draws visibly in the next animation frame. It helps the score only if this reduces blocking enough to outweigh the extra draw and any delay to visible content. It does not remove the work of drawing the map.

Lighthouse counts the part of each long task above 50 ms toward Total Blocking Time. Splitting work into shorter tasks therefore matters to this score. For example, one 100 ms task contributes 50 ms of blocking, while two separate 50 ms tasks contribute zero. This example uses the task durations on the timeline Lighthouse evaluates; it is not a manual conversion of our local traces. [Lighthouse's TBT explanation](https://developer.chrome.com/docs/lighthouse/performance/lighthouse-total-blocking-time).

TBT has a 30% weight in the saved reports. Largest Contentful Paint and the other loading metrics also affect the score, so reducing blocking does not guarantee a higher overall score. [Lighthouse's scoring explanation](https://developer.chrome.com/docs/lighthouse/performance/performance-scoring).

Each cell below is a median, unchanged → off-screen draw, from five runs per version. The medians are calculated separately: the run with the median score is not necessarily the run with median blocking time. TBT and LCP are Lighthouse's simulated metrics. They are not times read from the warm-up marks.

| Test | Score | TBT, ms | LCP, ms | Higher / lower / equal scores |
| --- | --- | --- | --- | --- |
| 1 | 44 → 48 | 1,012 → 679 | 6,793 → 6,758 | 2 / 2 / 1 |
| 2 | 44 → 48 | 1,165 → 842 | 6,786 → 6,745 | 4 / 1 / 0 |
| 3 | 56 → 48 | 848 → 729 | 6,834 → 6,768 | 2 / 2 / 1 |

Median blocking time fell in all three comparisons, but the overall median score fell in the third. Individual runs also changed substantially. In round 4 of the third comparison, for example, the unchanged page scored 70 with LCP of 2.89 seconds, while the changed page scored 40 with LCP of 6.78 seconds. Its blocking time was also higher in that round. Lower median blocking time alone therefore does not establish an overall score benefit.

These reports show whether the total score and blocking time improved; they do not establish that a particular driver pipeline was reused. The code adds a draw and delays visible rendering by one animation frame. A higher PageSpeed score is therefore not evidence by itself that the complete map appears sooner. The measurements apply to this initial-load prototype on the tested example.

`draw-metric-analysis.json` preserves every report's metric scores and their weighted contributions. It also retains the observed map timing marks, labelled separately from Lighthouse's simulated metrics.

</details>

<details>
<summary>One hidden 1 by 1 draw during loading: every hosted run</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 53 | 39 | 85 | 62 | 62 | 62 |
| After: one change | 🔴 41 | 🟢 43 | 🔴 60 | 🔴 52 | 🟢 84 | 🔴 52 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 513 | 1,514 | 244 | 313 | 318 | 318 |
| After: one change | 🔴 1,108 | 🟢 1,341 | 🔴 280 | 🔴 517 | 🟢 269 | 🔴 517 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-gcc-richards-publishers-peterson-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/i8sl2q6ipv?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-gcc-richards-publishers-peterson-trycloudflare-com-current-warmdraw-maplibre-gl-js-docs-examples-disable-map-rotation/i9aku98ltd?form_factor=mobile).

</details>

<details>
<summary>One full-size off-screen draw during loading: every hosted run</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 42 | 43 | 44 | 77 | 49 | 44 |
| After: one change | 🟢 52 | 🟢 51 | ⚪ 44 | 🔴 42 | 🔴 48 | 🟢 48 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,026 | 1,012 | 1,135 | 506 | 769 | 1,012 |
| After: one change | 🟢 642 | 🟢 568 | 🔴 1,266 | 🔴 1,088 | 🟢 679 | 🟢 679 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-gcc-richards-publishers-peterson-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/k6b5qlutb2?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-gcc-richards-publishers-peterson-trycloudflare-com-current-offscreen-maplibre-gl-js-docs-examples-disable-map-rotation/hecd5q99gk?form_factor=mobile).

</details>

<details>
<summary>Second test: one full-size off-screen draw during loading: every hosted run</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 44 | 44 | 38 | 41 | 66 | 44 |
| After: one change | 🟢 81 | 🟢 48 | 🟢 68 | 🟢 47 | 🔴 42 | 🟢 48 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,154 | 933 | 1,697 | 1,165 | 1,252 | 1,165 |
| After: one change | 🟢 386 | 🟢 842 | 🟢 988 | 🟢 721 | 🔴 1,387 | 🟢 842 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-gcc-richards-publishers-peterson-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/q638qxilth?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-gcc-richards-publishers-peterson-trycloudflare-com-current-offscreen-maplibre-gl-js-docs-examples-disable-map-rotation/vt7ua2gctw?form_factor=mobile).

</details>

<details>
<summary>Third test: one full-size off-screen draw during loading: every hosted run</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 43 | 37 | 56 | 70 | 84 | 56 |
| After: one change | 🔴 42 | 🟢 48 | ⚪ 56 | 🔴 40 | 🟢 85 | 🔴 48 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,033 | 2,008 | 392 | 848 | 314 | 848 |
| After: one change | 🔴 1,434 | 🟢 729 | 🟢 386 | 🔴 1,273 | 🟢 229 | 🟢 729 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-cheap-neck-springs-halloween-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/1h2rmw5g2q?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-cheap-neck-springs-halloween-trycloudflare-com-current-offscreen-maplibre-gl-js-docs-examples-disable-map-rotation/eccol5mrf4?form_factor=mobile).

</details>


<details>
<summary>Delay tile repaints by 100 ms: every hosted run</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 41 | 44 | 38 | 49 | 75 | 44 |
| After: one change | 🟢 51 | 🟢 52 | 🟢 42 | 🔴 40 | 🔴 54 | 🟢 51 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,148 | 940 | 1,639 | 673 | 604 | 940 |
| After: one change | 🟢 705 | 🟢 520 | 🟢 1,144 | 🔴 1,410 | 🟢 469 | 🟢 705 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/5cdrzt78nk?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-batch-maplibre-gl-js-docs-examples-disable-map-rotation/wbkm5t8nj5?form_factor=mobile).

</details>

<details>
<summary>Delay tile repaints by 100 ms, repeat</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 49 | 50 | 58 | 51 | 51 | 51 |
| After: one change | ⚪ 49 | 🔴 44 | 🔴 48 | 🔴 43 | 🔴 45 | 🔴 45 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 780 | 597 | 418 | 584 | 658 | 597 |
| After: one change | 🟢 630 | 🔴 979 | 🔴 858 | 🔴 1,285 | 🔴 836 | 🔴 858 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/cgryzx1oc6?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-batch-maplibre-gl-js-docs-examples-disable-map-rotation/i7qcgmb9zu?form_factor=mobile).

</details>

<details>
<summary>Disable fades: every hosted run</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 39 | 49 | 36 | 42 | 57 | 42 |
| After: one change | 🟢 44 | 🔴 44 | 🟢 52 | 🟢 66 | 🔴 53 | 🟢 52 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,429 | 634 | 2,277 | 1,079 | 434 | 1,079 |
| After: one change | 🟢 919 | 🔴 1,222 | 🟢 544 | 🔴 1,166 | 🔴 483 | 🟢 919 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/g25wnvl3bx?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-nofade-maplibre-gl-js-docs-examples-disable-map-rotation/1ryy9xmgau?form_factor=mobile).

</details>

<details>
<summary>Disable fades, repeat</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 39 | 49 | 40 | 45 | 51 | 45 |
| After: one change | 🟢 50 | 🟢 57 | 🔴 39 | 🔴 38 | 🔴 33 | 🔴 39 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,418 | 664 | 1,357 | 802 | 677 | 802 |
| After: one change | 🟢 762 | 🟢 350 | 🔴 1,454 | 🔴 1,602 | 🔴 3,145 | 🔴 1,454 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/226x01v516?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-nofade-maplibre-gl-js-docs-examples-disable-map-rotation/bo25fkduuk?form_factor=mobile).

</details>

<details>
<summary>Preload the style JSON: every hosted run</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 50 | 37 | 55 | 45 | 42 | 45 |
| After: one change | 🟢 57 | 🟢 47 | 🔴 43 | 🟢 76 | 🟢 44 | 🟢 47 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 487 | 1,801 | 511 | 715 | 1,052 | 715 |
| After: one change | 🔴 519 | 🟢 853 | 🔴 1,156 | 🟢 547 | 🔴 1,365 | 🔴 853 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/3v7bakcz5s?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-preload-maplibre-gl-js-docs-examples-disable-map-rotation/y66buupe4h?form_factor=mobile).

</details>

<details>
<summary>Preload the style JSON, repeat</summary>

PageSpeed performance score, higher is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 45 | 42 | 65 | 48 | 66 | 48 |
| After: one change | 🟢 47 | 🟢 46 | 🔴 48 | 🔴 40 | 🔴 49 | 🔴 47 |

Total Blocking Time in milliseconds, lower is better.

| Version | Round 1 | Round 2 | Round 3 | Round 4 | Round 5 | Median |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Before: unchanged | 1,070 | 1,399 | 1,260 | 859 | 1,184 | 1,184 |
| After: one change | 🟢 793 | 🟢 877 | 🟢 763 | 🔴 1,615 | 🟢 705 | 🟢 793 |

Example reports: [unchanged](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/ef6ghhq8d9?form_factor=mobile) and [with the change](https://pagespeed.web.dev/analysis/https-obtaining-alumni-msgid-browse-trycloudflare-com-current-preload-maplibre-gl-js-docs-examples-disable-map-rotation/kpdng3lgir?form_factor=mobile).

</details>

<details>
<summary>Library patches and the earlier combined comparison</summary>

I tested copies of the full documentation example using MapLibre 6.10.0, with the same host, page content, map style, and production build process. I compared the mobile PageSpeed score and Total Blocking Time, or TBT. TBT contributes to the PageSpeed score and measures blocking from long main-thread tasks during page loading. It therefore helps check whether moving shader work out of the first draws benefits the score.

The prewarm proposal aims to start shader preparation while tiles load, leaving less work for the first draws. The prototype splits program creation into two parts. It schedules compilation and linking after the style's first update, one program per task. It reads the link result and attribute and uniform locations when a draw first needs the program.

A table maps layer types to shader programs and their paint-property configurations. Programs already in the cache are reused. Programs that the table misses still use the existing draw path. This is a larger experimental change, and the mapping adds maintenance work. The prototype defers synchronous queries; it does not guarantee that they are free when the first draw reaches them.

A separate instrumented run confirmed that the standalone prototype linked five programs before the draws that used them. It also linked two programs that were not used before the map became idle. The unchanged version linked eight programs and used all eight. This checks that prewarming took place; it does not identify the cause of the PageSpeed score differences. The diagnostic run is excluded from the score tables.

The sky change avoids preparing and drawing a pass that adds no visible color. I tested skipping `drawSky` when both evaluated sky and horizon colors are fully transparent:

```ts
if (sky.properties.get('sky-color').a === 0 && sky.properties.get('horizon-color').a === 0) return;
```

This goes at the beginning of `src/webgl/draw/draw_sky.ts`, before program lookup, mesh setup, and drawing. It leaves the separate atmosphere draw unchanged.

The earlier hosted series tested the sky change alone and then added prewarm to check for an additional benefit. That combined prototype retained a disabled repaint-batching hook. The new standalone prewarm build removes that hook entirely and leaves sky drawing unchanged. The new comparison tests prewarm on its own.

Each row below contains five runs. The score lists are in round order:

| Library change | Scores | Median score | Median TBT |
| --- | --- | ---: | ---: |
| Unchanged | 64, 53, 41, 40, 40 | 41 | 1,245 ms |
| Skip transparent sky | 43, 58, 37, 50, 51 | 50 | 624 ms |
| Sky change plus prewarm | 48, 49, 51, 45, 38 | 48 | 704 ms |

Example reports: [unchanged library, score 41](https://pagespeed.web.dev/analysis/https-encountered-pest-category-northeast-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/ifieyz6itw?form_factor=mobile) and [sky change, score 50](https://pagespeed.web.dev/analysis/https-encountered-pest-category-northeast-trycloudflare-com-current-sky-maplibre-gl-js-docs-examples-disable-map-rotation/1btf6xxdr9?form_factor=mobile).

The lower median TBT makes the sky change interesting. However, the unchanged library itself scored anywhere from 40 to 64. The sky change scored higher in three rounds and lower in two. These results do not show that the nine-point median gain will repeat reliably.

This earlier combined comparison measures the effect of adding prewarm to the sky change. It does not isolate prewarm alone. The standalone comparison above fills that gap. These experiments do not measure the potential benefit of a smaller bundle or different worker loading.

</details>

<details>
<summary>How to interpret the blocking-time results</summary>

Total Blocking Time, or TBT, counts the portion of each long main-thread task above 50 ms during Lighthouse's measurement window. A 100 ms task contributes 50 ms; two separate 50 ms tasks contribute zero. These examples use task durations on the timeline Lighthouse evaluates. This explains why reducing work inside long tasks, or splitting work into separate tasks, is relevant to PageSpeed. [Lighthouse's TBT explanation](https://developer.chrome.com/docs/lighthouse/performance/lighthouse-total-blocking-time).

Lighthouse also simulates the page load under slower CPU and network conditions. Multiplying every task in a local trace by four and subtracting 50 ms does not reproduce that simulation or its measurement window. The reported scores and TBT values in the new comparisons come directly from Lighthouse, with no manual multiplier. [Lighthouse's throttling documentation](https://github.com/GoogleChrome/lighthouse/blob/main/docs/throttling.md).

TBT contributes 30% of the performance score in the saved hosted reports. LCP, FCP, Speed Index, and layout shifts also contribute, so reducing TBT does not translate directly into the same percentage improvement in the score. This investigation concerns the PageSpeed lab performance score, not a measurement of real visitors' Core Web Vitals.

</details>

<details>
<summary>Other local Lighthouse comparisons</summary>

These tests used production bundles based on MapLibre 6.9.0, Lighthouse 12.6.1, and Chrome for Testing 151. Each variant had five runs in rotating order, with a fresh browser for each run. These are separate from the hosted PageSpeed comparisons above.

The repaint experiment waited up to 100 ms before requesting a repaint for an arriving tile while its source was still loading. The aim was to group updates. Its control used the same hook with zero delay. Disabling fades aimed to avoid animation frames during loading. Preloading the style JSON aimed to start the next request in the loading chain sooner.

| Series | Variant | Median score | Median TBT |
| --- | --- | ---: | ---: |
| Live map assets | Unchanged | 83 | 531 ms |
| Live map assets | Repaint hook, zero delay | 82 | 554 ms |
| Live map assets | Repaint hook, 100 ms delay | 83 | 536 ms |
| Live map assets | `fadeDuration: 0` | 84 | 514 ms |
| Live map assets | Prewarm alone | 85 | 442 ms |
| Live map assets | Sky change alone | 86 | 417 ms |
| Saved map assets | Unchanged | 84 | 492 ms |
| Saved map assets | Sky change | 87 | 384 ms |
| Style preload comparison | Unchanged | 83 | 559 ms |
| Style preload comparison | Preload style JSON | 78 | 747 ms |
| Combined comparison | Unchanged | 84 | 533 ms |
| Combined comparison | Sky change | 89 | 364 ms |
| Combined comparison | Prewarm | 86 | 446 ms |
| Combined comparison | Sky change plus prewarm | 90 | 318 ms |

Compare variants within their own series. The local sky and prewarm results motivated the hosted tests. The earlier hosted tests did not establish the same additional benefit from combining them. The new standalone hosted comparisons above test each option separately. Repaint batching and disabling fades showed small median differences with overlapping results. Style preloading made the score worse in its comparison.

The saved-asset comparison removed external map requests as a source of variation. It still showed lower TBT with the sky change, but not a faster map load: median observed load was 276 ms unchanged and 339 ms with the patch. Median observed idle was 391 and 376 ms. Both had the same median simulated FCP and LCP. Lower TBT and earlier map completion are different results.

</details>

<details>
<summary>How the documentation page affected the score</summary>

I added this check during the investigation; it was not a CSS change proposed in this thread. The issue uses the documentation example as a PageSpeed test page. Its surrounding content therefore matters when interpreting the score. PageSpeed identified the heading or source-code block as the largest visible content. Its appearance determines Largest Contentful Paint, or LCP, which contributes to the score. The page's CSS requests delayed that rendering.

I replaced the parent page's four stylesheet links with `<style>` elements containing the saved Material main CSS, palette CSS, `extra.css`, and Google Fonts CSS. Font files stayed external, and the MapLibre iframe's CSS stayed unchanged. The map remained present. This is a website change, not an improvement that other sites get by updating MapLibre.

Each cell below summarizes five runs:

| Version | First set: median score / TBT / LCP | Repeat: median score / TBT / LCP |
| --- | --- | --- |
| Unchanged | 48 / 708 ms / 6.83 s | 43 / 993 ms / 6.75 s |
| CSS in HTML | 84 / 486 ms / 2.10 s | 80 / 664 ms / 2.10 s |
| CSS plus sky change | 49 / 961 ms / 6.42 s | 78 / 737 ms / 2.13 s |

Across both sets, CSS alone scored 70–90 versus 40–54 for its controls. CSS plus the sky change scored 47–92, so adding the library change did not produce a consistent further improvement.

A separate test made only the Google Fonts stylesheet non-blocking, using preload followed by stylesheet activation. Its five scores were 37, 40, 51, 60, and 42, with median TBT 1,053 ms. Its control was the first hosted series, with median score 41. This smaller website change did not establish an improvement.

Inlining all CSS repeats the styles in each HTML page instead of sharing separately cached stylesheets between page visits. The tested change demonstrates an effect on this example; adopting it in the documentation build is a separate decision.

Example PageSpeed reports: [unchanged page](https://pagespeed.web.dev/analysis/https-encountered-pest-category-northeast-trycloudflare-com-current-maplibre-gl-js-docs-examples-disable-map-rotation/xbx2i84z3j?form_factor=mobile) and [CSS in HTML](https://pagespeed.web.dev/analysis/https-encountered-pest-category-northeast-trycloudflare-com-current-inline-maplibre-gl-js-docs-examples-disable-map-rotation/wqmssam1a0?form_factor=mobile).

</details>

<details>
<summary>Earlier traces and shader experiments, with corrected interpretation</summary>

The earlier September 14 PageSpeed runs of the live 6.9.0 documentation example scored 39, 37, and 39, with TBT of 1,277, 1,656, and 1,325 ms. The reports listed long tasks from both MapLibre and the documentation site's own JavaScript. Those reports did not isolate the share caused by each component.

The local trace experiments used a development build. In the frame-cadence and repaint tests, prewarm was already enabled in both the control and variants. They were not measurements of an unchanged production bundle.

The earlier hand estimate was `sum(max(0, 4 * taskDuration - 50))`, applied to recorded main-thread tasks. It produced 774 and 811 ms in two traces where Lighthouse itself reported TBT of 349 and 356 ms. Both Lighthouse runs scored 50. That difference is why the hand estimates must not be presented as PageSpeed TBT.

The frame tasks in those traces included both JavaScript and rendering waits. They accounted for 92–97% of the hand estimate, but that does not mean JavaScript accounted for only 3–8%. It also does not establish a limit on the benefit of changing JavaScript or bundle size.

The corrected frame descriptions were an initial background/sky frame of about 45–52 ms, a later frame containing tile geometry and labels of about 105–144 ms, and later substantial loading frames around 29–38 ms. The first two were not separate tile-only and label-only frames.

The earlier fade and batching comparisons reduced the hand estimate of frame-related blocking:

- Disabling fades changed the range from about 770–940 ms to 730–860 ms.
- A 100 ms repaint delay changed it from about 800–870 ms to 670–740 ms.

The total number of animation-frame tasks did not decrease consistently. These results prompted the production-bundle Lighthouse checks above; they were not measured PageSpeed score gains.

The local renderer was SwiftShader, and `KHR_parallel_shader_compile` was unavailable. One trace showed a 143 ms main-thread `Commit` alongside about 140 ms in the GPU process's `readPixels` path, including `ContextVk::finishImpl`. This identifies a local rendering wait. It does not prove that the same wait accounts for all of Google's reported blocking time.

A separate CARTO Positron testbed used a 412 by 288 map, 93 style layers, and seven shader programs:

- Repainting a loaded map took about 13 ms. Clearing the program cache and repainting took about 110 ms. Those tasks contained roughly 3 versus 28 ms of JavaScript and 10 versus 82 ms in `Commit`. This points to first-use program and draw setup as useful areas to investigate; it does not isolate every part of the extra work.
- The final prewarm prototype's first tile frame took 87 ms, within three baseline runs of 87, 96, and 98 ms. Its JavaScript portion was 18 ms versus about 21–23 ms in the baseline. One prototype run did not establish a repeatable gain.
- An earlier synchronous prewarm version produced 58–116 ms prewarm tasks while the page animated the camera. Moving the queries into prewarm therefore did not necessarily remove long tasks.
- Warm-up draws using zero-sized, 1 by 1, and 4 by 4 viewports performed about 29–33 ms of GPU-process work, followed by about 80 ms during the later real draw. They did not eliminate the expensive draw. The test did not establish exactly which driver pipelines were reused.
- Rendering a full frame into an off-screen framebuffer first left about 103 ms of GPU-process work at the visible repaint, compared with 99 ms without it. The visible repaint's JavaScript fell to about 3 ms, but the rendering wait remained.
- Loading the page twice in one browser still produced main-thread `Commit` times of about 145 and 140 ms. That does not prove that the browser cached no shader data, and it is not a cold-load PageSpeed comparison.

The earlier sky experiment measured six steady-state repaints per case, rather than page loading. With and without the sky pass, the medians were 8.3/8.5 ms for small Positron, 9.5/7.8 ms for full-height Positron, and 27.6/27.8 ms for full-height demotiles. Those mixed repaint results do not answer whether skipping the initial sky setup improves PageSpeed, which is why the later load tests were needed.

The [earlier outputs](https://gist.github.com/martinfrancois/df78ec8087152ec431c3eeb28aec17a5), [scripts](https://gist.github.com/martinfrancois/21587fb06eaf4f891b62cfbb6013b983), and [prototype patch](https://gist.github.com/martinfrancois/682aebf46bf150d5c66a5be96fb3e1cd) remain available. Their old labels describing the hand estimate as PageSpeed's calculation are superseded by the explanation above. These gists contain the earlier work, not the new hosted comparisons.

</details>

<details>
<summary>Exact setup, checks, and how to reproduce</summary>

All archives, hashes, and download instructions are available in the [reproduction gist](https://gist.github.com/martinfrancois/4f07a011f83d532d0119586c2aecb613).

The earlier reproduction files are split between [pagespeed-evidence.tar.gz](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/pagespeed-evidence.tar.gz) and [pagespeed-standalone-reports.tar.gz](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/pagespeed-standalone-reports.tar.gz). Extract these two archives into the same directory. The new initial-load draw comparisons have their own self-contained [pagespeed-draw-evidence.tar.gz](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/pagespeed-draw-evidence.tar.gz) archive, described above. The archives contain the measured bundles, patches, captured documentation assets, local test pages, scripts with a dependency lockfile, raw reports, report links, screenshots, and a SHA-256 manifest. [REPRODUCE.md](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/REPRODUCE.md) provides the full commands. The bundle preserves the older scripts separately and explains their measurement limitations. Optional [pagespeed-traces.tar.gz](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/pagespeed-traces.tar.gz), [pagespeed-standalone-screenshots-1.tar.gz](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/pagespeed-standalone-screenshots-1.tar.gz) and [pagespeed-standalone-screenshots-2.tar.gz](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/pagespeed-standalone-screenshots-2.tar.gz) archives contain the retained local traces and hosted report screenshots. All three required archives together contain the raw Lighthouse reports needed to verify every table.

The source bases were:

- Local production comparisons: 6.9.0, commit `21864b4537b2169507e58e4610d0adcd28ccef8a`.
- Hosted comparisons: 6.10.0, commit `d1934699ed3159f1e5d689078dd2e677f3ed34e3`.

The 6.10.0 builds used that tag's own lockfile and dependencies. The unmodified main, shared, and worker bundles matched the official release byte for byte. The bundle also supplies the exact older measured files, whose builds reused dependencies from the existing 6.9.0 checkout.

The live-map comparisons used `https://demotiles.maplibre.org/style.json`, center `[-122.65, 45.52]`, and zoom 3. The local page filled a 412 by 823 viewport at DPR 1.75. The hosted page retained the documentation layout and its 400 CSS-pixel-high iframe.

Local tests used Node 24.20.0, Chrome for Testing 151.0.7922.173, Lighthouse 12.6.1, and Lighthouse-managed mobile simulation with CPU multiplier four. Production assets used precomputed gzip. The fixed-asset series served saved styles, tiles, and glyphs locally.

The September 18 hosted reports used Lighthouse 13.4.1 and Chrome 153.0.8010.36. Their recorded CPU multiplier was 1.2, not four. The Google workers' CPU benchmark indices varied. Runs used the normal PageSpeed web interface; each report's JSON records its complete settings. Hosted pages used live map assets. Within each comparison, all variants used the same temporary HTTPS host. Cloudflare added its email-protection script to the hosted pages; every new report includes the same 1,239-byte script. These were matched copies, not a deployment to maplibre.org.

The earlier archives contain 70 completed local reports, 130 hosted comparison reports, and one retained hosted report of the original live page, which scored 40. The draw archive adds 40 hosted comparison reports, for 241 retained reports in total. The 80 new hosted reports cover standalone comparisons and their repeats. Keep each comparison with its own control; do not combine scores from different pages or series.

After extracting the bundle:

```bash
cd pagespeed-evidence
python3 verify-evidence.py
npm ci
npx playwright install chromium
```

Verification checks file hashes, the saved scores and TBT values, HTTP error responses, and map load/idle marks. To serve the full documentation experiment:

```bash
python3 public-server-standalone.py
```

Expose port 8140 through a public HTTPS host. For example, run this in another terminal:

```bash
cloudflared tunnel --url http://127.0.0.1:8140
```

Use the resulting HTTPS origin in place of `PUBLIC_ORIGIN`:

```bash
node psi-series.mjs PUBLIC_ORIGIN current,current-prewarm 5 verify-prewarm-only
python3 summarize-hosted.py verify-prewarm-only
node psi-series.mjs PUBLIC_ORIGIN current,current-batch 5 verify-batch-only
python3 summarize-hosted.py verify-batch-only
node psi-series.mjs PUBLIC_ORIGIN current,current-batch 5 verify-batch-confirm
python3 summarize-hosted.py verify-batch-confirm
node psi-series.mjs PUBLIC_ORIGIN current,current-nofade 5 verify-nofade-only
python3 summarize-hosted.py verify-nofade-only
node psi-series.mjs PUBLIC_ORIGIN current,current-nofade 5 verify-nofade-confirm
python3 summarize-hosted.py verify-nofade-confirm
node psi-series.mjs PUBLIC_ORIGIN current,current-sky 5 verify-sky-confirm
python3 summarize-hosted.py verify-sky-confirm
node psi-series.mjs PUBLIC_ORIGIN current,current-preload 5 verify-preload-only
python3 summarize-hosted.py verify-preload-only
node psi-series.mjs PUBLIC_ORIGIN current,current-preload 5 verify-preload-confirm
python3 summarize-hosted.py verify-preload-confirm
```

Run the comparisons one after another. The runner alternates order within each pair, requests fresh mobile audits, saves the full reports, and retains every completed score. [REPRODUCE.md](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/REPRODUCE.md) also includes the commands for the earlier sky, combined, and CSS comparisons. Use a new prefix for a new comparison. The temporary hosts have been stopped; the saved reports remain available. Google controls the version used for future hosted audits, so compare the variants within the new run rather than expecting identical scores.

For local comparisons, run `python3 server.py` in another terminal, then:

```bash
python3 measure.py baseline,batch0,batch,nofade,prewarm,nosky 5 verify-local
python3 summarize.py verify-local
python3 measure.py baseline,nosky 5 fixture-verify
python3 summarize.py fixture-verify
python3 measure.py baseline,preload 5 verify-preload
python3 summarize.py verify-preload
python3 measure.py baseline,nosky,prewarm,combined 5 verify-combined
python3 summarize.py verify-combined
```

The bundle includes the exact measured production files. To rebuild the hosted variants from source instead:

```bash
python3 rebuild.py current current-prewarm current-batch current-sky current-combined
```

This fetches the recorded commit, applies each patch in its own checkout, installs dependencies, runs the code-generation and production-build steps, and compares the output with the measured bundles. It does not overwrite those supplied bundles. [REPRODUCE.md](https://raw.githubusercontent.com/martinfrancois/maplibre-gl-js/46f345daa555b996c92816ccd249ce00048348b1/research/pagespeed-1876/REPRODUCE.md) also covers rebuilding the older variants and running the historical development testbed.

The new standalone builds passed 86 existing tests for prewarm and 96 for batching, plus TypeScript checking for both. The option checks confirmed `fadeDuration: 0`, the 100 ms batching setting, and one early style preload request with no duplicate style download. All 16 served files for the unchanged, prewarm, batching, and sky builds matched their recorded bundle hashes.

Every completed comparison report recorded map load and idle, with no HTTP error responses. Map screenshots matched their controls. The CSS-only full-page screenshot also matched. The combined CSS/sky screenshot differed along the source-tab underline, with no changed map pixels. In the new standalone checks, prewarm, batching, and no-fade full-page screenshots matched the unchanged page. The preload screenshot differed by 72 color channels at the source-tab underline; the map matched. The new sky check had the same small underline difference, with matching map pixels.

Ten additional sky scenarios matched, including transparent and visible sky/horizon, pitch, roll, globe, and settled transitions in both directions. The sky patch passed 26 existing tests and TypeScript checking. These checks help ensure an improved score was not caused by missing content; they do not cover every possible style or every transition frame.

Uncompressed setup runs were excluded. One local run failed when its server exited. One earlier hosted attempt and one attempt in the new fade repeat produced no report within 180 seconds. These failed attempts were retained and retried. No completed score was discarded.

AI assistance: Claude Code and Codex helped run the experiments, prepare the prototypes, and draft this comment under my direction.

</details>

**Comment by HarelM** [comment]
I'm exhausted of reading AI generated monologues, I didn't finished reading your post. Can you summarize this in two short sentences please?

**Comment by martinfrancois** [comment]
@HarelM AI helped with the investigation and write-up, but the amount of detail was my choice so others could build on the tests and know what I tried exactly. I've now added an overview table and put the supporting details in expandable sections.

The summary you asked for: the library changes I tried out including all suggestions mentioned in this issue and some other things did not consistently improve PageSpeed scores, even when blocking time fell in some cases.
